packages/flutter_tools/lib/src/platform_step.dart

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'android/apk.dart';
+import 'base/context.dart';
+import 'build_info.dart';
+import 'cache.dart';
+import 'project.dart';
+
+/// The [PlatformBuilders] instance.
+PlatformBuilders get platformBuilders => context.get<PlatformBuilders>();
+
+/// A registry that selects the correct [PlatformBuilder] based on the provided
+/// build info.
+class PlatformBuilders {
+  /// Create a new [PlatformBuilder] with `_platformSteps` as platform specific
+  /// implementations.
+  const PlatformBuilders([this._platformSteps = const <PlatformBuildStep>[
+    AndroidPlatformBuildStep(),
+  ]]);
+
+  final List<PlatformBuildStep> _platformSteps;
+
+  /// Build the platform specific bundle for [flutterProject].
+  ///
+  /// Selects the [PlatformBuildStep] that supports the required [TargetPlatform]
+  /// and [BuildMode] requested.
+  ///
+  /// Throws an [AmbiguousPlatformBuilder] if multiple build steps
+  /// match.
+  ///
+  /// Throws a [MissingPlatformBuilder] if no build steps match.
+  PlatformBuildStep selectPlatform({
```

Comment:
How about "selectBuildStep"

Thinking ahead, this method may in the future also need to understand a concept of the stage of the build. E.g. something -> source, source -> kernel, kernel -> kernel, kernel -> native, kernel -> something platform specific.

Reply:
This sounds like the right direction, but our current implementation of most native builds is somewhat backwards. For example, on Android we invoke gradle which re-entrantly invokes flutter build apk.

In this structure, the platform build comes first, and contains all of the other build steps, then returns to the platform step.